### PR TITLE
Adding option to create a partially heterogeneous assemblies for 1D XS model

### PR DESCRIPTION
**Pull request by mgjarrett**
## What is the change?

The `latticePhysicsInterface` and `crossSectionGroupManager` modules are being updated to enable a new type of 1D cylindrical cross section model. The key features added are:

- New type of block collection in `crossSectionGroupManager`, `CylindricalComponentsPartHetAverageBlockCollection`, which extends the existing `CylindricalComponentsAverageBlockCollection` with a modified method for calculating the average nuclide temperatures within the representative block.
- New `crossSectionControlSettings`:
  1. `mergeIntoFuel` allows a user to specify components to be merged into the fuel for the cross section model
  2. `ductHeterogeneous` enables the user to select the partially heterogeneous 1D model (as opposed to fully heterogeneous)
  3. `traceIsotopeThreshold` allows a user to split low-concentration isotopes out of the 1D cross section model and instead run a 0D model to obtain cross sections for those isotopes. 

Much of the implementation details related to these new `crossSectionControl` options would need to be implemented in the downstream plugins that extend the `latticePhysicsInterface`.

## Why is the change being made?

These features are being added to enable downstream lattice physics plugins to have more options and flexibility when it comes to building a 1D cylindrical equivalent model of a 2D hexagonal assembly for cross section generation. Since a fully-heterogeneous 1D cylindrical model can pose a significantly more compute-intensive problem for the lattice physics solver than a simple 0D model, it is helpful to have options for simplifying the 1D model. Combining small/unimportant components into others (`mergeIntoFuel`), homogenizing several regions together (`ductHeterogeneous`), and pulling out trace isotopes that don't contribute significantly to reaction rates (`traceIsotopeThreshold`) can all help to reduce the complexity of the 1D model.

---

## Checklist

- [x] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [x] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify any new/changed code.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/developer/tooling.html#add-release-notes) have been updated if necessary.
- [x] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/nuclearDataIO/tests/test_xsLibraries.py [hunk 1]
```diff
@@ -127,7 +127,7 @@ def test_mergeFailsWithNonIsotxsFiles(self):
             os.remove(dummyFileName)
 
         with TemporaryDirectoryChanger():
-            dummyFileName = "ISOtopics.txt"
+            dummyFileName = "ISO[]"
```
Comment (jakehader) [inline armi/nuclearDataIO/tests/test_xsLibraries.py:130]: What's the reason for this change?

**Comment by jakehader** [comment]
@mgjarrett - Do you have some images you could share on what this model looks like when enabled? Is there any part of the user manual / documentation should be amended to describe this in more detail with examples?

File: armi/nuclearDataIO/tests/test_xsLibraries.py (hunk 1 above)
Comment (mgjarrett) [reply armi/nuclearDataIO/tests/test_xsLibraries.py:130]: I wanted to allow files to have lower-case letters, e.g. `ISOAa` or `ISOBa`. This was previously not recognized as an ISOTXS file.

**Comment by mgjarrett** [comment]
> @mgjarrett - Do you have some images you could share on what this model looks like when enabled? Is there any part of the user manual / documentation should be amended to describe this in more detail with examples?

Here are two PNGs generated by ARMI of the converted 1D cylindrical representation of the 2D hexagon. This is for a hexagon with 10 rings of pins (271 total pins in the lattice).

Partially Heterogeneous (everything inside the duct is homogenized):
1D_CYL_PART_HET_AVG_AA-cyl (1)

Fully heterogeneous:
1D_CYL_AVG_AA-cyl

File: armi/physics/neutronics/crossSectionSettings.py [hunk 2]
```diff
@@ -57,6 +58,8 @@
 CONF_COMPONENT_AVERAGING = "averageByComponent"
 CONF_XS_MAX_ATOM_NUMBER = "xsMaxAtomNumber"
 CONF_MIN_DRIVER_DENSITY = "minDriverDensity"
+CONF_PARTIALLY_HETEROGENEOUS = "partiallyHeterogeneous"
```
Comment (zachmprince) [inline armi/physics/neutronics/crossSectionSettings.py:61]: I personally think the name of this setting should be more descriptive like `ductHeterogeneous`.

File: armi/reactor/converters/blockConverters.py [hunk 3]
```diff
@@ -645,6 +661,25 @@ def _classifyComponents(self):
 
         return list(sorted(pinComponents)), nonPins
 
+    def _buildInsideDuct(self):
+        """Build a homogenized material of the components inside the duct."""
+        blockType = self._sourceBlock.getType()
+        blockName = f"Homogenized {blockType}"
+        newBlock, mixtureFlags = stripComponents(self._sourceBlock, Flags.DUCT)
+        outerDiam = getOuterDiamFromIDAndArea(0.0, newBlock.getArea())
+        circle = components.Circle(
```
Comment (zachmprince) [inline armi/reactor/converters/blockConverters.py:670]: Is this really supposed to be a circle? Or should it be derived from the shape of the duct, i.e. hexagon?

File: armi/reactor/converters/blockConverters.py [hunk 4]
```diff
@@ -824,3 +862,61 @@ def radiiFromRingOfRods(distToRodCenter, numRods, rodRadii, layout="hexagon"):
         rLast, bigRLast = rodRadius, distFromCenterComp
 
     return sorted(radiiFromRodCenter)
+
+
+def stripComponents(block, compFlags):
```
Comment (zachmprince) [inline armi/reactor/converters/blockConverters.py:867]: Would a more accurate neme be `mergeInnerComponents`?

File: armi/reactor/converters/blockConverters.py (hunk 3 above)
Comment (zachmprince) [reply armi/reactor/converters/blockConverters.py:670]: Removing this one since I was being dumb.

File: armi/reactor/converters/blockConverters.py (hunk 4 above)
Comment (mgjarrett) [reply armi/reactor/converters/blockConverters.py:867]: This function doesn't actually homogenize the components, it just creates a new block and pulls off the components that aren't going to be homogenized, and then returns that block.

**Review by zachmprince — APPROVED** [review]

**Comment by john-science** [comment]
bump

File: armi/physics/neutronics/crossSectionGroupManager.py [hunk 5]
```diff
@@ -648,6 +653,64 @@ def _getNucTempHelper(self):
         return nvt, nv
 
 
+class CylindricalComponentsDuctHetAverageBlockCollection(
```
Comment (john-science) [inline armi/physics/neutronics/crossSectionGroupManager.py:656]: Honest question: why is this class in ARMI?

Is it just because `BLOCK_COLLECTIONS` is used in `blockCollectionFactory()`?

File: armi/physics/neutronics/crossSectionSettings.py [hunk 6]
```diff
@@ -429,10 +444,31 @@ class XSModelingOptions:
         (e.g., fission products) as a depletion product of an isotope with a much
         smaller atomic number.
 
+    averageByComponent: bool
```
Comment (john-science) [inline armi/physics/neutronics/crossSectionSettings.py:447]: And the new settings just have to go in ARMI, because they're so closely tied to the current ARMI XSGM?

File: armi/physics/neutronics/crossSectionGroupManager.py (hunk 5 above)
Comment (mgjarrett) [reply armi/physics/neutronics/crossSectionGroupManager.py:656]: Yes, pretty much. It's just an extension of the `CylindricalComponentsAverageBlockCollection` with modified behavior in a few circumstances.

File: armi/physics/neutronics/crossSectionSettings.py (hunk 6 above)
Comment (mgjarrett) [reply armi/physics/neutronics/crossSectionSettings.py:447]: `averageByComponent` was already in ARMI, it was just missing from the docstring.

The new settings `ductHeterogeneous` and `traceIsotopeThreshold` are here because they are new `XSModelingOptions`. There is no downstream extension of `XSModelingOptions`, it's all handled at this level.

`ductHeterogeneous` is also used as an argument passed to `blockConverters.HexComponentsToCylConverter`.

**Review by john-science — APPROVED** [review]

File: doc/release/0.4.rst [hunk 7]
```diff
@@ -25,6 +25,7 @@ New Features
 #. Provide ``Block.getInputHeight`` for determining the height of a block from blueprints. (`PR#1927 <https://github.com/terrapower/armi/pull/1927>`_)
 #. Improve performance by changing the lattice physics interface so that cross sections are not updated on ``everyNode`` calls during coupled calculations (`PR#1963 <https://github.com/terrapower/armi/pull/1963>`_)
 #. Improve efficiency of reaction rate calculations. (`PR#1887 <https://github.com/terrapower/armi/pull/1887>`_)
+#. Add some new options for simplification of 1D cross section modeling. (`PR#1949 <https://github.com/terrapower/armi/pull/1949>`_)
```
Comment (john-science) [inline doc/release/0.4.rst:28]:
```suggestion
#. Adding new options for simplifying 1D cross section modeling. (`PR#1949 <https://github.com/terrapower/armi/pull/1949>`_)
```
